src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -73,6 +73,8 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument)
                 builder.Insert(i++, @using);
             }
 
+            InheritImportedDirectives(document, codeDocument.Source.FilePath);
```

Comment:
So I experimented with reversing the lowering order so I could conditionally not add a directive instead of removing it later. Problem with this approach is it totally messes with other directives like `@inject` who implement their own import behavior (it makes the most relevant directive first instead of last).

Therefore, went with an inner-pass approach where we remove directives that have been added if we feel that they should not have been imported.